<a href="https://colab.research.google.com/github/nicolemichaud03/Recipe-Recommender-System/blob/main/NNnotebook.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Neural Network Recipe Recommendation System with Embeddings for Dietary Restriction

### Business Problem:


### Pre-Processing and Data Exploration:

Loading the data and necessary packages:

In [1]:
# TF's recommender imports
!pip install -q tensorflow-recommenders
!pip install -q tensorflow_ranking
!pip install -q --upgrade tensorflow-datasets
!pip install -q scann

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 475.2/475.2 MB 2.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.5/5.5 MB 93.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 442.0/442.0 kB 44.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 83.9 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
scann 1.2.10 requires tensorflow~=2.13.0, but you have tensorflow 2.15.0.post1 which is incompatible.
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
tensorflow-serving-api 2.14.1 requires tensorflow<3,>=2.14.1, but you have tensorflow 2.13.1 which is incompatible.


In [2]:
import tensorflow as tf
print("TensorFlow version:", tf.__version__)

TensorFlow version: 2.13.1


In [3]:
import pandas as pd
import nltk
nltk.download('punkt')
from nltk.corpus import stopwords
import re
import numpy as np
import pickle
from nltk.tokenize import RegexpTokenizer, word_tokenize
import io
from collections import defaultdict
import os
import pprint
import tempfile
from typing import Dict, Text
import numpy as np
import tensorflow as tf
import tensorflow_datasets as tfds
import tensorflow_recommenders as tfrs
import tensorflow_ranking as tfr

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


In [4]:
# from google.colab import drive
# drive.mount('/content/drive')

Mounted at /content/drive


Loading the user and item data and viewing it:

In [4]:
#upload the kaggle.json file to load kaggle data
from google.colab import files
files.upload()

Saving kaggle.json to kaggle (1).json


{'kaggle (1).json': b'{"username":"nicolemichaud","key":"e7fb46517234dcc36218486df964d630"}'}

In [5]:
!mkdir -p ~/.kaggle
!cp kaggle.json ~/.kaggle/

In [6]:
!pip install -q kaggle

# This permissions change avoids a warning on Kaggle tool startup.
!chmod 600 ~/.kaggle/kaggle.json

In [7]:
#download the data from kaggle
!kaggle datasets download -d shuyangli94/food-com-recipes-and-user-interactions

100% 267M/267M [00:07<00:00, 41.7MB/s]
100% 267M/267M [00:07<00:00, 39.4MB/s]


In [ ]:
#unzip data from kaggle
import zipfile

# Define the path to your zip file
file_path = '/content/drive/MyDrive/Colab Notebooks/capstone_data/food-com-recipes-and-user-interactions.zip'  # Replace 'your_file.zip' with your file's name

# Unzip the file to a specific destination
with zipfile.ZipFile(file_path, 'r') as zip_ref:
    zip_ref.extractall('/content/drive/MyDrive/Colab Notebooks/capstone_data')  # Replace 'destination_folder' with your desired folder


In [ ]:
#read in the specific datasets to be used:
user_data = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/capstone_data/RAW_interactions.csv')
recipe_data = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/capstone_data/RAW_recipes.csv')

In [ ]:
user_data.head()

In [9]:
user_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1132367 entries, 0 to 1132366
Data columns (total 5 columns):
 #   Column     Non-Null Count    Dtype 
---  ------     --------------    ----- 
 0   user_id    1132367 non-null  int64 
 1   recipe_id  1132367 non-null  int64 
 2   date       1132367 non-null  object
 3   rating     1132367 non-null  int64 
 4   review     1132198 non-null  object
dtypes: int64(3), object(2)
memory usage: 43.2+ MB


In [10]:

recipe_data.head()

name      id  minutes  \
0  arriba   baked winter squash mexican style  137739       55   
1            a bit different  breakfast pizza   31490       30   
2                   all in the kitchen  chili  112140      130   
3                          alouette  potatoes   59389       45   
4          amish  tomato ketchup  for canning   44061      190   

   contributor_id   submitted  \
0           47892  2005-09-16   
1           26278  2002-06-17   
2          196586  2005-02-25   
3           68585  2003-04-14   
4           41706  2002-10-25   

                                                tags  \
0  ['60-minutes-or-less', 'time-to-make', 'course...   
1  ['30-minutes-or-less', 'time-to-make', 'course...   
2  ['time-to-make', 'course', 'preparation', 'mai...   
3  ['60-minutes-or-less', 'time-to-make', 'course...   
4  ['weeknight', 'time-to-make', 'course', 'main-...   

                                    nutrition  n_steps  \
0       [51.5, 0.0, 13.0, 0.0, 2.0, 0.0, 4.0]       11   
1   [173.4, 18.0, 0.0, 17.0, 22.0, 35.0, 1.0]        9   
2  [269.8, 22.0, 32.0, 48.0, 39.0, 27.0, 5.0]        6   
3   [368.1, 17.0, 10.0, 2.0, 14.0, 8.0, 20.0]       11   
4   [352.9, 1.0, 337.0, 23.0, 3.0, 0.0, 28.0]        5   

                                               steps  \
0  ['make a choice and proceed with recipe', 'dep...   
1  ['preheat oven to 425 degrees f', 'press dough...   
2  ['brown ground beef in large pot', 'add choppe...   
3  ['place potatoes in a large pot of lightly sal...   
4  ['mix all ingredients& boil for 2 1 / 2 hours ...   

                                         description  \
0  autumn is my favorite time of year to cook! th...   
1  this recipe calls for the crust to be prebaked...   
2  this modified version of 'mom's' chili was a h...   
3  this is a super easy, great tasting, make ahea...   
4  my dh's amish mother raised him on this recipe...   

                                         ingredients  n_ingredients  
0  ['winter squash', 'mexican seasoning', 'mixed ...              7  
1  ['prepared pizza crust', 'sausage patty', 'egg...              6  
2  ['ground beef', 'yellow onions', 'diced tomato...             13  
3  ['spreadable cheese with garlic and herbs', 'n...             11  
4  ['tomato juice', 'apple cider vinegar', 'sugar...              8

In [11]:
recipe_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 231637 entries, 0 to 231636
Data columns (total 12 columns):
 #   Column          Non-Null Count   Dtype 
---  ------          --------------   ----- 
 0   name            231636 non-null  object
 1   id              231637 non-null  int64 
 2   minutes         231637 non-null  int64 
 3   contributor_id  231637 non-null  int64 
 4   submitted       231637 non-null  object
 5   tags            231637 non-null  object
 6   nutrition       231637 non-null  object
 7   n_steps         231637 non-null  int64 
 8   steps           231637 non-null  object
 9   description     226658 non-null  object
 10  ingredients     231637 non-null  object
 11  n_ingredients   231637 non-null  int64 
dtypes: int64(5), object(7)
memory usage: 21.2+ MB


In [12]:
recipe_data = recipe_data.rename(columns={"id": "recipe_id"})

In [13]:
print(len(user_data['user_id'].unique()))
print(len(recipe_data['recipe_id'].unique()))

226570
231637


With 226,570 users and 231,637 recipes, there are less users than there are recipes. Therefore, it is probably best for our recommender system to be user-user based.

Dropping the columns I know I won't be working with:

In [14]:
user_recipe_ratings = user_data.drop(columns=['review', 'date'])

In [15]:
recipe_data = recipe_data.drop(columns=['contributor_id', 'submitted', 'nutrition', 'steps', 'minutes', 'n_steps', 'description', 'n_ingredients'])
recipe_data = recipe_data.rename(columns={"id": "recipe_id"})

In [16]:
# Converting id values to strings to be compatible for
# later use with tensorflow
user_data['recipe_id'] = user_data['recipe_id'].astype(str)
recipe_data['recipe_id'] = recipe_data['recipe_id'].astype(str)
user_data['user_id'] = user_data['user_id'].astype(str)

In [17]:
# Making sure text features are strings so that they can be
# cleaned appropriately
recipe_data['name'] = recipe_data['name'].astype(str)
#recipe_data['description'] = recipe_data['description'].astype(str)
recipe_data['tags'] = recipe_data['tags'].astype(str)

In [18]:

# Creating a function to perform cleaning steps at once
# (Removes numbers and unnecessary characters,
# makes all letters lowercase, removes stopwords)
nltk.download('stopwords')
stopwords_list = stopwords.words('english')

no_bad_chars = re.compile('[!\"#$%&()*+-./:;<=>?@[\]^_`{|}~\n - ]')
no_nums = re.compile('[\d-]')

def clean_text(text):
    text = no_nums.sub('', text)
    text = no_bad_chars.sub(' ', text)
    text = text.lower()
    text = ' '.join(word for word in text.split() if word not in stopwords_list)
    return text

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


In [19]:
#Applying text cleaning function to text columns
recipe_data_cleaned = recipe_data.copy()
recipe_data_cleaned['name'] = (recipe_data['name']).apply(clean_text)
#recipe_data_cleaned['description'] = (recipe_data['description']).apply(clean_text)
recipe_data_cleaned['tags'] = (recipe_data['tags']).apply(clean_text)
recipe_data_cleaned.head()

name recipe_id  \
0  arriba baked winter squash mexican style    137739   
1             bit different breakfast pizza     31490   
2                             kitchen chili    112140   
3                         alouette potatoes     59389   
4              amish tomato ketchup canning     44061   

                                                tags  \
0  'minutesorless' 'timetomake' 'course' 'maining...   
1  'minutesorless' 'timetomake' 'course' 'maining...   
2  'timetomake' 'course' 'preparation' 'maindish'...   
3  'minutesorless' 'timetomake' 'course' 'maining...   
4  'weeknight' 'timetomake' 'course' 'mainingredi...   

                                         ingredients  
0  ['winter squash', 'mexican seasoning', 'mixed ...  
1  ['prepared pizza crust', 'sausage patty', 'egg...  
2  ['ground beef', 'yellow onions', 'diced tomato...  
3  ['spreadable cheese with garlic and herbs', 'n...  
4  ['tomato juice', 'apple cider vinegar', 'sugar...

Feature Engineering to categorize each recipe as different diet types (Vegetarian, Vegan, and/or Gluten-Free):

In [20]:
# creating a new column to classify recipes as gluten-free or not
GF = []
#tags column contains the most information on gluten-free recipes
#(see miscellaneous notebook)
for row in recipe_data_cleaned['tags']:
    if "gluten-free" in row : GF.append("Gluten-Free")
    elif "gluten free" in row : GF.append("Gluten-Free")
    else: GF.append("None")

In [21]:
recipe_data_cleaned['GF'] = GF

In [22]:
#Ingredient lists for diet filtering:
vegan = ['ham', 'beef', 'meat', 'chicken', 'pork', 'bacon', 'sausage', 'lamb', 'veal', 'turkey', 'steak', 'rib', 'frankfurter', 'duck', 'poultry', 'goat', 'liver', 'hen', 'quail', 'brisket', 'goose','fish', 'shrimp', 'seafood', 'crab', 'lobster', 'clam', 'oyster', 'scallop', 'mussel', 'cod', 'salmon', 'halibut', 'shellfish', 'roe', 'tuna', 'caviar', 'pollock', 'yellowtail', 'squid', 'calamari', 'octopus', 'crawfish', 'crayfish', 'sardine', 'trout', 'flounder', 'anchovy', 'bass', 'haddock', 'sole','egg', 'honey','milk', 'cheese', 'yogurt', 'mayonnaise', 'butter', 'margarine', 'cream']

vegetarian = ['ham', 'beef', 'meat', 'chicken', 'pork', 'bacon', 'sausage', 'lamb', 'veal', 'turkey', 'steak', 'rib', 'frankfurter', 'duck', 'poultry', 'goat', 'liver', 'hen', 'quail', 'brisket', 'goose','fish', 'shrimp', 'seafood', 'crab', 'lobster', 'clam', 'oyster', 'scallop', 'mussel', 'cod', 'salmon', 'halibut', 'shellfish', 'roe', 'tuna', 'caviar', 'pollock', 'yellowtail', 'squid', 'calamari', 'octopus', 'crawfish', 'crayfish', 'sardine', 'trout', 'flounder', 'anchovy', 'bass', 'haddock', 'sole']

In [23]:
# creating two new columns to classify recipes as vegetarian or not
# and as vegan or not:

recipe_data_cleaned['vegetarian'] = None
recipe_data_cleaned['vegan'] = None

In [24]:
# Filtering through the 'ingedients' column for ingredients that
# aren' t vegetarian or vegan
vege_pattern = '|'.join(vegetarian)
vegan_pattern = '|'.join(vegan)


recipe_data_cleaned.vegetarian = recipe_data_cleaned.ingredients.str.contains(vege_pattern)
recipe_data_cleaned.vegan = recipe_data_cleaned.ingredients.str.contains(vegan_pattern)

In [25]:
# Changing Boolean values to 0 and 1 for easy counting later
recipe_data_cleaned['vegetarian'] = recipe_data_cleaned['vegetarian'].astype(str)
recipe_data_cleaned['vegetarian'] = recipe_data_cleaned['vegetarian'].replace({'False': 'Vegetarian', 'True': 'None'})

In [26]:
recipe_data_cleaned['vegan'] = recipe_data_cleaned['vegan'].astype(str)
recipe_data_cleaned['vegan'] = recipe_data_cleaned['vegan'].replace({'False': 'Vegan', 'True': 'None'})

In [27]:
#Making column names match and merging dfs to classify user diets based on the recipes they've used
recipe_data_cleaned = recipe_data_cleaned.rename(columns = {'id': 'recipe_id'})

#user_diets = pd.merge(user_data, recipe_data_cleaned, on='recipe_id', how='left')

In [28]:
#user_diets.head()

In [29]:
recipe_data_cleaned['diets_combined'] = recipe_data_cleaned[['vegetarian', 'vegan', 'GF']].values.tolist()
recipe_data_cleaned = recipe_data_cleaned.drop(columns=['GF', 'vegetarian', 'vegan'])
recipe_data_cleaned.head()

name recipe_id  \
0  arriba baked winter squash mexican style    137739   
1             bit different breakfast pizza     31490   
2                             kitchen chili    112140   
3                         alouette potatoes     59389   
4              amish tomato ketchup canning     44061   

                                                tags  \
0  'minutesorless' 'timetomake' 'course' 'maining...   
1  'minutesorless' 'timetomake' 'course' 'maining...   
2  'timetomake' 'course' 'preparation' 'maindish'...   
3  'minutesorless' 'timetomake' 'course' 'maining...   
4  'weeknight' 'timetomake' 'course' 'mainingredi...   

                                         ingredients  \
0  ['winter squash', 'mexican seasoning', 'mixed ...   
1  ['prepared pizza crust', 'sausage patty', 'egg...   
2  ['ground beef', 'yellow onions', 'diced tomato...   
3  ['spreadable cheese with garlic and herbs', 'n...   
4  ['tomato juice', 'apple cider vinegar', 'sugar...   

              diets_combined  
0   [Vegetarian, None, None]  
1         [None, None, None]  
2         [None, None, None]  
3   [Vegetarian, None, None]  
4  [Vegetarian, Vegan, None]

In [30]:
# #combining the diet-classified recipe data with the user data based on which recipes each user has rated

# user_diets = (user_diets.groupby(['user_id']).agg({'recipe_id': lambda x: x.tolist(), 'rating': lambda x: x.tolist(), 'GF': sum, 'vegetarian':sum, 'vegan': sum}).reset_index())
# user_diets.head()

In [31]:
# # creating new columns to classify users as the different diet types
# user_diets['is_vegetarian'] = None
# user_diets['is_vegan'] = None
# user_diets['is_GF'] = None
# user_diets.head()

In [32]:
# #getting a total count of the recipes that each user has rated
# user_diets['recipe_totals'] = user_diets['recipe_id'].str.len()

In [33]:
# #classifying a user as vegetarian if at least 75% of the recipes they've rated are vegetarian
# user_diets['is_vegetarian'] = np.where(user_diets['vegetarian'] >= ((user_diets['recipe_totals'])*(.75)), 1, 0)


In [34]:
# #classifying a user as vegan if at least 75% of the recipes they've rated are vegan
# user_diets['is_vegan'] = np.where(user_diets['vegan'] >= ((user_diets['recipe_totals'])*(.75)), 1, 0)

In [35]:
# #classifying a user as gluten-free if at least 75% of the recipes they've rated are gluten-free
# user_diets['is_GF'] = np.where(user_diets['GF'] >= ((user_diets['recipe_totals'])*(.75)), 1, 0)

### Modeling

Before I can use the data with TensorFlow, I need to ensure that all features (except for rating) are strings:

In [36]:
recipe_data_cleaned['diets_combined'] = recipe_data_cleaned['diets_combined'].astype(str)
recipe_data_cleaned['recipe_id'] = recipe_data_cleaned['recipe_id'].astype(str)
recipe_data_cleaned['name'] = recipe_data_cleaned['name'].astype(str)
user_recipe_ratings['user_id'] = user_recipe_ratings['user_id'].astype(str)
user_recipe_ratings['recipe_id'] = user_recipe_ratings['recipe_id'].astype(str)

In [37]:
user_recipe_ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1132367 entries, 0 to 1132366
Data columns (total 3 columns):
 #   Column     Non-Null Count    Dtype 
---  ------     --------------    ----- 
 0   user_id    1132367 non-null  object
 1   recipe_id  1132367 non-null  object
 2   rating     1132367 non-null  int64 
dtypes: int64(1), object(2)
memory usage: 25.9+ MB


Next, I create one dataframe with all the necessary features to make things simpler:

In [38]:
# Merge movie data
merged_df = user_recipe_ratings.merge(recipe_data_cleaned, on="recipe_id", how="left")
merged_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1132367 entries, 0 to 1132366
Data columns (total 7 columns):
 #   Column          Non-Null Count    Dtype 
---  ------          --------------    ----- 
 0   user_id         1132367 non-null  object
 1   recipe_id       1132367 non-null  object
 2   rating          1132367 non-null  int64 
 3   name            1132367 non-null  object
 4   tags            1132367 non-null  object
 5   ingredients     1132367 non-null  object
 6   diets_combined  1132367 non-null  object
dtypes: int64(1), object(6)
memory usage: 69.1+ MB


This merged dataframe needs to be turned into a TensorFlow dataset:

In [39]:
merged_ds = tf.data.Dataset.from_tensor_slices(dict(merged_df))

recipes_ds = tf.data.Dataset.prefetch(merged_ds, buffer_size=tf.data.AUTOTUNE)


Before I create any models, I need to map the correct data features so that they can be easily called when I write the model functions:

In [41]:
#preparing the data

# Select the basic features.
user_ratings = recipes_ds.map(lambda x: {
    "user_id": x["user_id"],
    "rating": x["rating"],
    "recipe_id": x["recipe_id"],
    "diets_combined": x["diets_combined"],
    "name": x["name"]})


#recipe_map = recipes_ds.map(lambda x:x["recipe_id"])

In [42]:
#making lists of all values in each column
recipe_ids = user_ratings.map(lambda x: x["recipe_id"])
user_ids = user_ratings.map(lambda x: x["user_id"])
names = user_ratings.map(lambda x: x["name"])
diets = user_ratings.map(lambda x: x["diets_combined"])


In [43]:
#making lists of only the unique values in each column (this will be the vocabularies for the models)
unique_recipe_ids = np.unique(np.concatenate(list(recipe_ids)))
unique_user_ids = np.unique(np.concatenate(list(user_ids)))
unique_recipe_names = np.unique(np.concatenate(list(names)))
unique_diets = np.unique(np.concatenate(list(diets)))

In [46]:
#making sure each unique value list is in string format
unique_user_ids = unique_user_ids.astype(str)
unique_recipe_ids = unique_recipe_ids.astype(str)
unique_recipe_names = unique_recipe_names.astype(str)
unique_diets = unique_diets.astype(str)


In [46]:
# This is multitask recommender model adapted from TensorFlow's website. It conducts both retrieval and ranking tasks depending
# on which weight you assign each task. This model contains no extra feature embeddings (only looks at user_id, recipe_id, and ratings for creating recommendations)

class UserRecipesModel(tfrs.models.Model):

  def __init__(self, rating_weight: float, retrieval_weight: float) -> None:
    # We take the loss weights in the constructor: this allows us to instantiate
    # several model objects with different loss weights.

    super().__init__()

    embedding_dimension = 32

    # User and recipe models.
    self.recipe_model: tf.keras.layers.Layer = tf.keras.Sequential([
      tf.keras.layers.StringLookup(
        vocabulary=unique_recipe_ids, mask_token=None),
      tf.keras.layers.Embedding(len(unique_recipe_ids) + 1, embedding_dimension)
    ])
    self.user_model: tf.keras.layers.Layer = tf.keras.Sequential([
      tf.keras.layers.StringLookup(
        vocabulary=unique_user_ids, mask_token=None),
      tf.keras.layers.Embedding(len(unique_user_ids) + 1, embedding_dimension)
    ])

    # A small model to take in user and recipe embeddings and predict ratings.
    # We can make this as complicated as we want as long as we output a scalar
    # as our prediction.
    self.rating_model = tf.keras.Sequential([
        tf.keras.layers.Dense(256, activation="relu"),
        tf.keras.layers.Dense(128, activation="relu"),
        tf.keras.layers.Dense(1),
    ])

    # The tasks.
    self.rating_task: tf.keras.layers.Layer = tfrs.tasks.Ranking(
        loss=tf.keras.losses.MeanSquaredError(),
        metrics=[tf.keras.metrics.RootMeanSquaredError()],
    )
    self.retrieval_task: tf.keras.layers.Layer = tfrs.tasks.Retrieval(
        metrics=tfrs.metrics.FactorizedTopK(
            candidates=recipes.batch(128).map(self.recipe_model)
        )
    )


    # "since we have two tasks and two losses - we need to decide on how important each loss is.
    # We can do this by giving each of the losses a weight, and treating these weights as hyperparameters"

    # The loss weights.
    self.rating_weight = rating_weight
    self.retrieval_weight = retrieval_weight

  def call(self, features: Dict[Text, tf.Tensor]) -> tf.Tensor:
    # We pick out the user features and pass them into the user model.
    user_embeddings = self.user_model(features["user_id"])
    # And pick out the recipe features and pass them into the recipe model.
    recipe_embeddings = self.recipe_model(features["recipe_id"])

    return (
        user_embeddings,
        recipe_embeddings,
        # We apply the multi-layered rating model to a concatentation of
        # user and recipe embeddings.
        self.rating_model(
            tf.concat([user_embeddings, recipe_embeddings], axis=1)
        ),
    )

  def compute_loss(self, features: Dict[Text, tf.Tensor], training=False) -> tf.Tensor:

    ratings = features.pop("rating")

    user_embeddings, recipe_embeddings, rating_predictions = self(features)

    # We compute the loss for each task.
    rating_loss = self.rating_task(
        labels=ratings,
        predictions=rating_predictions,
    )
    retrieval_loss = self.retrieval_task(user_embeddings, recipe_embeddings)

    # And combine them using the loss weights.
    return (self.rating_weight * rating_loss
            + self.retrieval_weight * retrieval_loss)

In [47]:
#Ranking specialized model (only the ranking task has weight)
model = UserRecipesModel(rating_weight=1.0, retrieval_weight=0.0)
model.compile(optimizer=tf.keras.optimizers.Adagrad(0.1))

In [48]:
#Then shuffle, batch, and cache the training and evaluation data:
tf.random.set_seed(42)
shuffled = recipes_ds.shuffle(100_000, seed=42, reshuffle_each_iteration=False)

train = shuffled.take(80_000)
test_val = shuffled.skip(80_000).take(20_000)
validation = test_val.skip(10_000).take(10_000)
test = test_val.skip(10_000).take(10_000)

cached_train = train.shuffle(100_000).batch(8192)
cached_val = validation.batch(4096).cache()
cached_test = test.batch(4096).cache()


In [49]:
model.fit(cached_train, epochs=3)
metrics = model.evaluate(cached_val, return_dict=True)

print(f"Retrieval top-100 accuracy: {metrics['factorized_top_k/top_100_categorical_accuracy']:.3f}.")
print(f"Ranking RMSE: {metrics['root_mean_squared_error']:.3f}.")

The RMSE of this Ranking-only model was 1.267.

Retrieval top-100 accuracy: 0.001

In [50]:
#Retrieval specialized model (only the retrieval task has weight)
model = UserRecipesModel(rating_weight=0.0, retrieval_weight=1.0)
model.compile(optimizer=tf.keras.optimizers.Adagrad(0.1))

In [51]:
model.fit(cached_train, epochs=3)
metrics = model.evaluate(cached_val, return_dict=True)

print(f"Retrieval top-100 accuracy: {metrics['factorized_top_k/top_100_categorical_accuracy']:.3f}.")
print(f"Ranking RMSE: {metrics['root_mean_squared_error']:.3f}.")

The RMSE of this Retrieval-only model was: 4.528

Retrieval top-100 accuracy: 0.036

In [52]:
#joint model (both tasks have weight)
model = UserRecipesModel(rating_weight=1.0, retrieval_weight=1.0)
model.compile(optimizer=tf.keras.optimizers.Adagrad(0.1))

In [53]:
model.fit(cached_train, epochs=3)
metrics = model.evaluate(cached_val, return_dict=True)

print(f"Retrieval top-100 accuracy: {metrics['factorized_top_k/top_100_categorical_accuracy']:.3f}.")
print(f"Ranking RMSE: {metrics['root_mean_squared_error']:.3f}.")

The RMSE of this combined (ranking and retrieval) model was: 1.342

Retrieval top-100 accuracy: 0.037

In [54]:
# #make predictions
# trained_recipe_embeddings, trained_user_embeddings, predicted_rating = model({
#       #insert specific user and item IDs here:
#       "user_id": np.array([" "]),
#       "recipe_id": np.array([" "])
#   })
# print("Predicted rating:")
# print(predicted_rating)

In [55]:
#recipe_ds_plus = tf.data.Dataset.from_tensor_slices(dict(recipe_data_cleaned))

#### Model with extra embeddings:

The next model has the same embeddings as the original model, but with two additional ones: a text embedding for the recipe names feature, and a text embedding for the recipe diet types feature that I created.

In [48]:
#the user model with no additional embeddings:
class UserModel2(tf.keras.Model):
    def __init__(self):
        super().__init__()
        self.user_embeddings = tf.keras.Sequential(
[tf.keras.layers.StringLookup(vocabulary=unique_user_ids,    mask_token=None),tf.keras.layers.Embedding(len(unique_user_ids)+1, 32)])
    def call(self, inputs):
          return self.user_embeddings(inputs["user_id"])

In [74]:
#The recipe model with the additional name and diet embeddings:
class RecipeModel2(tf.keras.Model):

  def __init__(self):
    super().__init__()

    max_tokens = 10_000

    self.recipe_embedding = tf.keras.Sequential([
      tf.keras.Sequential(
          [tf.keras.layers.StringLookup(vocabulary=unique_recipe_ids, mask_token=None),
           tf.keras.layers.Embedding(len(unique_recipe_ids)+1, 32)
           ])
      ])

    self.name_text_embedding = tf.keras.Sequential([
        tf.keras.layers.StringLookup(vocabulary=unique_recipe_names, mask_token=None),
        tf.keras.layers.Embedding(len(unique_recipe_names)+1, 32)])

    self.diet_embedding = tf.keras.Sequential([
        tf.keras.layers.StringLookup(vocabulary=unique_diets, mask_token=None),
        tf.keras.layers.Embedding(len(unique_diets)+1, 32)])

    self.text_vectorizer =  tf.keras.layers.TextVectorization(max_tokens=max_tokens)

    self.name_text_embedding = tf.keras.Sequential([
        self.text_vectorizer,
        tf.keras.layers.Embedding(max_tokens, 32, mask_zero=True),
        tf.keras.layers.GlobalAveragePooling1D()])

    self.text_vectorizer.adapt(names)
    self.text_vectorizer.adapt(diets)

  def call(self, inputs):
    return tf.concat( [self.recipe_embedding(inputs),  self.name_text_embedding(inputs), self.diet_embedding(inputs)],axis=1)


In [76]:
#Model combining both the user and recipe models, assigning the tasks, and computing the loss
#https://blog.searce.com/recommendation-systems-using-tensorflow-recommenders-d7d12167b0b7
class RecipeRecommendModel(tfrs.models.Model):
    def __init__(self, rating_weight, retrieval_weight):
        super().__init__()
        embedding_dimension = 32
        self.query_model = tf.keras.Sequential([UserModel2(), tf.keras.layers.Dense(embedding_dimension)])
        self.candidate_model = tf.keras.Sequential([RecipeModel2(), tf.keras.layers.Dense(embedding_dimension)])
        self.rating_model = tf.keras.Sequential([tf.keras.layers.Dense(256, activation="relu"), tf.keras.layers.Dense(128, activation="relu"), tf.keras.layers.Dense(1)])
        self.retrieval_task = tfrs.tasks.Retrieval(metrics=tfrs.metrics.FactorizedTopK(candidates=recipe_map.batch(128).map(self.candidate_model)))
        self.rating_task = tfrs.tasks.Ranking(loss=tf.keras.losses.MeanSquaredError(), metrics=[tf.keras.metrics.RootMeanSquaredError()])
       # The loss weights.
        self.rating_weight = rating_weight
        self.retrieval_weight = retrieval_weight

    def call(self, features: Dict[Text, tf.Tensor]) -> tf.Tensor:
        user_embeddings = self.query_model({"user_id": features["user_id"]})
        recipe_embeddings = self.candidate_model({"recipe_id":features["recipe_id"]})
        return (user_embeddings, recipe_embeddings, self.rating_model(tf.concat([user_embeddings, recipe_embeddings],axis=1)))

    def compute_loss(self, features: Dict[Text, tf.Tensor], training=False) -> tf.Tensor:

        ratings = features.pop("rating")
        user_embeddings, recipe_embeddings, rating_predictions = self(features)
        # We compute the loss for each task.
        rating_loss = self.rating_task(labels=ratings, predictions=rating_predictions)
        retrieval_loss = self.retrieval_task(user_embeddings, recipe_embeddings)
        # And combine them using the loss weights.
        return (self.rating_weight * rating_loss + self.retrieval_weight * retrieval_loss)

In [80]:
model_2 = RecipeRecommendModel(1, 1)
model_2.compile(optimizer=tf.keras.optimizers.Adagrad(learning_rate=0.5))


KeyboardInterrupt: 

In [81]:
# tf.random.set_seed(42)
# shuffled = recipes_ds.shuffle(100_000, seed=42, reshuffle_each_iteration=False)

# train_2 = shuffled.take(80_000)
# test_2 = shuffled.skip(80_000).take(20_000)

# cached_train_2 = train_2.shuffle(100_000).batch(2048)
# cached_test_2 = test_2.batch(4096).cache()


In [83]:
model_2.fit(cached_train, epochs=3)
metrics = model_2.evaluate(cached_val, return_dict=True)

print(f"Retrieval top-100 accuracy: {metrics['factorized_top_k/top_100_categorical_accuracy']:.3f}.")
print(f"Ranking RMSE: {metrics['root_mean_squared_error']:.3f}.")

5/5 [==============================] - 566s 110s/step - factorized_top_k/top_1_categorical_accuracy: 1.0000e-04 - factorized_top_k/top_5_categorical_accuracy: 1.0000e-04 - factorized_top_k/top_10_categorical_accuracy: 1.5000e-04 - factorized_top_k/top_50_categorical_accuracy: 2.0000e-04 - factorized_top_k/top_100_categorical_accuracy: 3.5000e-04 - root_mean_squared_error: 61.4468 - loss: 3301005.0417 - regularization_loss: 0.0000e+00 - total_loss: 3301005.0417
Retrieval top-100 accuracy: 0.000.
Ranking RMSE: 61.447.


This multi-task (retrieval and ranking) model with the extra embeddings of recipe name and the diet-type features I added had a ranking RMSE of about 61.447 and a top-100 accuracy of 0.000.

In [59]:
# extra embeddings model with only the ranking task
# model_3 = RecipeRecommendModel(1, 0)
# model_3.compile(optimizer=tf.keras.optimizers.Adagrad(learning_rate=0.5))

In [60]:
# model_3.fit(cached_train, epochs=3)

# metrics_3 = model_3.evaluate(cached_val, return_dict=True)

# print(f"Retrieval top-100 accuracy: {metrics_3['factorized_top_k/top_100_categorical_accuracy']:.3f}.")
# print(f"Ranking RMSE: {metrics_3['root_mean_squared_error']:.3f}.")

performed much worse with only the ranking task. RMSE = 440.949, 0.000 top-100 accuracy.

In [ ]:
#Retrieval Recs:
scann = tfrs.layers.factorized_top_k.ScaNN(model.query_model, num_leaves=1000, num_leaves_to_search=100, num_reordering_candidates=100, k=15)
scann.index_from_dataset(tf.data.Dataset.zip((movie_map.batch(100), movie_map.batch(100).map(model.candidate_model))))
_, titles = scann({"user_id":tf.constant(["221"])})

In [ ]:
#Ranking the Recs:
res = {"movie_title":[],"rank":[]}
test_data = {"user_id":tf.constant(["221"])}
for data in titles:
    res["movie_title"].append(data.decode())
    test_data["movie_title"] = np.array([data.decode()])
    _, _, rating = model(test_data)
    res["rank"].append(int(rating[0][0]))

In [49]:
# recipe model with only the diet embedding added
class RecipeModel3(tf.keras.Model):

  def __init__(self):
    super().__init__()

    max_tokens = 10_000

    self.recipe_embedding = tf.keras.Sequential([
      tf.keras.Sequential(
          [tf.keras.layers.StringLookup(vocabulary=unique_recipe_ids, mask_token=None),
           tf.keras.layers.Embedding(len(unique_recipe_ids)+1, 32)
           ])
      ])

    # self.name_text_embedding = tf.keras.Sequential([
    #     tf.keras.layers.StringLookup(vocabulary=unique_recipe_names, mask_token=None),
    #     tf.keras.layers.Embedding(len(unique_recipe_names)+1, 32)])

    self.diet_embedding = tf.keras.Sequential([
        tf.keras.layers.StringLookup(vocabulary=unique_diets, mask_token=None),
        tf.keras.layers.Embedding(len(unique_diets)+1, 32)])

    self.text_vectorizer =  tf.keras.layers.TextVectorization(max_tokens=max_tokens)

    # self.name_text_embedding = tf.keras.Sequential([
    #     self.text_vectorizer,
    #     tf.keras.layers.Embedding(max_tokens, 32, mask_zero=True),
    #     tf.keras.layers.GlobalAveragePooling1D()])

    # self.text_vectorizer.adapt(names)
    self.text_vectorizer.adapt(diets)

  def call(self, inputs):
    return tf.concat( [self.recipe_embedding(inputs), self.diet_embedding(inputs)],axis=1)


In [50]:
#combined model )(diet feature but no name feature)

#https://blog.searce.com/recommendation-systems-using-tensorflow-recommenders-d7d12167b0b7
class RecipeRecommendModel2(tfrs.models.Model):
    def __init__(self, rating_weight, retrieval_weight):
        super().__init__()
        embedding_dimension = 32
        self.query_model = tf.keras.Sequential([UserModel2(), tf.keras.layers.Dense(embedding_dimension)])
        self.candidate_model = tf.keras.Sequential([RecipeModel3(), tf.keras.layers.Dense(embedding_dimension)])
        self.rating_model = tf.keras.Sequential([tf.keras.layers.Dense(256, activation="relu"), tf.keras.layers.Dense(128, activation="relu"), tf.keras.layers.Dense(1)])
        self.retrieval_task = tfrs.tasks.Retrieval(metrics=tfrs.metrics.FactorizedTopK(candidates=recipe_map.batch(128).map(self.candidate_model)))
        self.rating_task = tfrs.tasks.Ranking(loss=tf.keras.losses.MeanSquaredError(), metrics=[tf.keras.metrics.RootMeanSquaredError()])
       # The loss weights.
        self.rating_weight = rating_weight
        self.retrieval_weight = retrieval_weight

    def call(self, features: Dict[Text, tf.Tensor]) -> tf.Tensor:
        user_embeddings = self.query_model({"user_id": features["user_id"]})
        recipe_embeddings = self.candidate_model({"recipe_id":features["recipe_id"]})
        return (user_embeddings, recipe_embeddings, self.rating_model(tf.concat([user_embeddings, recipe_embeddings],axis=1)))

    def compute_loss(self, features: Dict[Text, tf.Tensor], training=False) -> tf.Tensor:

        ratings = features.pop("rating")
        user_embeddings, recipe_embeddings, rating_predictions = self(features)
        # We compute the loss for each task.
        rating_loss = self.rating_task(labels=ratings, predictions=rating_predictions)
        retrieval_loss = self.retrieval_task(user_embeddings, recipe_embeddings)
        # And combine them using the loss weights.
        return (self.rating_weight * rating_loss + self.retrieval_weight * retrieval_loss)

In [51]:
model_4 = RecipeRecommendModel2(1, 1)
model_4.compile(optimizer=tf.keras.optimizers.Adagrad(learning_rate=0.5))

In [52]:
# tf.random.set_seed(42)
# shuffled = recipes_ds.shuffle(100_000, seed=42, reshuffle_each_iteration=False)

# train_4 = shuffled.take(80_000)
# test_4 = shuffled.skip(80_000).take(20_000)

# cached_train_4 = train_4.shuffle(100_000).batch(2048)
# cached_test_4 = test_4.batch(4096).cache()


In [53]:
model_4.fit(cached_train, epochs=3)

Epoch 1/3
40/40 [==============================] - 2441s 61s/step - factorized_top_k/top_1_categorical_accuracy: 0.0180 - factorized_top_k/top_5_categorical_accuracy: 0.0180 - factorized_top_k/top_10_categorical_accuracy: 0.0180 - factorized_top_k/top_50_categorical_accuracy: 0.0184 - factorized_top_k/top_100_categorical_accuracy: 0.0187 - root_mean_squared_error: 4267.6567 - loss: 17573463.4296 - regularization_loss: 0.0000e+00 - total_loss: 17573463.4296
Epoch 2/3
40/40 [==============================] - 2718s 68s/step - factorized_top_k/top_1_categorical_accuracy: 5.0000e-05 - factorized_top_k/top_5_categorical_accuracy: 6.2500e-05 - factorized_top_k/top_10_categorical_accuracy: 1.5000e-04 - factorized_top_k/top_50_categorical_accuracy: 6.2500e-04 - factorized_top_k/top_100_categorical_accuracy: 0.0011 - root_mean_squared_error: 24.6618 - loss: 119588.4777 - regularization_loss: 0.0000e+00 - total_loss: 119588.4777
Epoch 3/3
40/40 [==============================] - 2732s 68s/step - 

In [54]:
metrics_4 = model_4.evaluate(cached_val, return_dict=True)

print(f"Retrieval top-100 accuracy: {metrics_4['factorized_top_k/top_100_categorical_accuracy']:.3f}.")
print(f"Ranking RMSE: {metrics_4['root_mean_squared_error']:.3f}.")

5/5 [==============================] - 618s 123s/step - factorized_top_k/top_1_categorical_accuracy: 0.0000e+00 - factorized_top_k/top_5_categorical_accuracy: 5.0000e-05 - factorized_top_k/top_10_categorical_accuracy: 1.0000e-04 - factorized_top_k/top_50_categorical_accuracy: 2.5000e-04 - factorized_top_k/top_100_categorical_accuracy: 7.5000e-04 - root_mean_squared_error: 8.3858 - loss: 117851.8411 - regularization_loss: 0.0000e+00 - total_loss: 117851.8411
Retrieval top-100 accuracy: 0.001.
Ranking RMSE: 8.386.


Model with just the diet-type extra feature had a top-100 Retrieval accuracy of 0.001, and an RMSE of 8.386.

In [55]:
# recipe model with only the name embedding added
class RecipeModel4(tf.keras.Model):

  def __init__(self):
    super().__init__()

    max_tokens = 10_000

    self.recipe_embedding = tf.keras.Sequential([
      tf.keras.Sequential(
          [tf.keras.layers.StringLookup(vocabulary=unique_recipe_ids, mask_token=None),
           tf.keras.layers.Embedding(len(unique_recipe_ids)+1, 32)
           ])
      ])

    self.name_text_embedding = tf.keras.Sequential([
        tf.keras.layers.StringLookup(vocabulary=unique_recipe_names, mask_token=None),
        tf.keras.layers.Embedding(len(unique_recipe_names)+1, 32)])

    # self.diet_embedding = tf.keras.Sequential([
    #     tf.keras.layers.StringLookup(vocabulary=unique_diets, mask_token=None),
    #     tf.keras.layers.Embedding(len(unique_diets)+1, 32)])

    self.text_vectorizer =  tf.keras.layers.TextVectorization(max_tokens=max_tokens)

    self.name_text_embedding = tf.keras.Sequential([
        self.text_vectorizer,
        tf.keras.layers.Embedding(max_tokens, 32, mask_zero=True),
        tf.keras.layers.GlobalAveragePooling1D()])

    self.text_vectorizer.adapt(names)
    # self.text_vectorizer.adapt(diets)

  def call(self, inputs):
    return tf.concat( [self.recipe_embedding(inputs), self.name_text_embedding(inputs)],axis=1)


In [56]:
#combined model )(diet feature but no name feature)
#https://blog.searce.com/recommendation-systems-using-tensorflow-recommenders-d7d12167b0b7
class RecipeRecommendModel3(tfrs.models.Model):
    def __init__(self, rating_weight, retrieval_weight):
        super().__init__()
        embedding_dimension = 32
        self.query_model = tf.keras.Sequential([UserModel2(), tf.keras.layers.Dense(embedding_dimension)])
        self.candidate_model = tf.keras.Sequential([RecipeModel4(), tf.keras.layers.Dense(embedding_dimension)])
        self.rating_model = tf.keras.Sequential([tf.keras.layers.Dense(256, activation="relu"), tf.keras.layers.Dense(128, activation="relu"), tf.keras.layers.Dense(1)])
        self.retrieval_task = tfrs.tasks.Retrieval(metrics=tfrs.metrics.FactorizedTopK(candidates=recipe_map.batch(128).map(self.candidate_model)))
        self.rating_task = tfrs.tasks.Ranking(loss=tf.keras.losses.MeanSquaredError(), metrics=[tf.keras.metrics.RootMeanSquaredError()])
       # The loss weights.
        self.rating_weight = rating_weight
        self.retrieval_weight = retrieval_weight

    def call(self, features: Dict[Text, tf.Tensor]) -> tf.Tensor:
        user_embeddings = self.query_model({"user_id": features["user_id"]})
        recipe_embeddings = self.candidate_model({"recipe_id":features["recipe_id"]})
        return (user_embeddings, recipe_embeddings, self.rating_model(tf.concat([user_embeddings, recipe_embeddings],axis=1)))

    def compute_loss(self, features: Dict[Text, tf.Tensor], training=False) -> tf.Tensor:

        ratings = features.pop("rating")
        user_embeddings, recipe_embeddings, rating_predictions = self(features)
        # We compute the loss for each task.
        rating_loss = self.rating_task(labels=ratings, predictions=rating_predictions)
        retrieval_loss = self.retrieval_task(user_embeddings, recipe_embeddings)
        # And combine them using the loss weights.
        return (self.rating_weight * rating_loss + self.retrieval_weight * retrieval_loss)

In [57]:
model_5 = RecipeRecommendModel2(1, 1)
model_5.compile(optimizer=tf.keras.optimizers.Adagrad(learning_rate=0.5))

In [58]:
model_5.fit(cached_train_4, epochs=3)
metrics_5 = model_5.evaluate(cached_test_4, return_dict=True)

print(f"Retrieval top-100 accuracy: {metrics_5['factorized_top_k/top_100_categorical_accuracy']:.3f}.")
print(f"Ranking RMSE: {metrics_5['root_mean_squared_error']:.3f}.")

Epoch 1/3
40/40 [==============================] - 2473s 62s/step - factorized_top_k/top_1_categorical_accuracy: 0.0243 - factorized_top_k/top_5_categorical_accuracy: 0.0243 - factorized_top_k/top_10_categorical_accuracy: 0.0243 - factorized_top_k/top_50_categorical_accuracy: 0.0249 - factorized_top_k/top_100_categorical_accuracy: 0.0270 - root_mean_squared_error: 3592.4275 - loss: 12704148.4777 - regularization_loss: 0.0000e+00 - total_loss: 12704148.4777
Epoch 2/3
40/40 [==============================] - 2657s 66s/step - factorized_top_k/top_1_categorical_accuracy: 1.0000e-04 - factorized_top_k/top_5_categorical_accuracy: 1.0000e-04 - factorized_top_k/top_10_categorical_accuracy: 2.1250e-04 - factorized_top_k/top_50_categorical_accuracy: 6.7500e-04 - factorized_top_k/top_100_categorical_accuracy: 0.0013 - root_mean_squared_error: 118.1342 - loss: 249612.2531 - regularization_loss: 0.0000e+00 - total_loss: 249612.2531
Epoch 3/3
5/5 [==============================] - 629s 125s/step - f

This model with only the additional embedding for the recipe names had a retrieval top-100 accuracy of 0.001 and an RMSE of 30.929.

### Evaluation

### Limitations and Next Steps